packages/trace-viewer/src/ui/codegen.ts

```diff
@@ -0,0 +1,305 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { Language } from '@isomorphic/locatorGenerators';
+import type * as har from '@trace/har';
+
+interface APIRequestCodegen {
+  generatePlaywrightRequestCall(request: har.Request, body: string | undefined): string;
+}
+
+class JSCodeGen implements APIRequestCodegen {
+  generatePlaywrightRequestCall(request: har.Request, body: string | undefined): string {
+    let method = request.method.toLowerCase();
+    const url = new URL(request.url);
+    const urlParam = `${url.origin}${url.pathname}`;
+    const options: any = {};
+    if (!['delete', 'get', 'head', 'post', 'put', 'patch'].includes(method)) {
+      options.method = method;
+      method = 'fetch';
+    }
+    if (url.searchParams.size)
+      options.params = Object.fromEntries(url.searchParams.entries());
+    if (body)
+      options.data = body;
+    if (request.headers.length)
+      options.headers = Object.fromEntries(request.headers.map(header => [header.name, header.value]));
+
+    const params = [`'${urlParam}'`];
+    const hasOptions = Object.keys(options).length > 0;
+    if (hasOptions)
+      params.push(this.prettyPrintObject(options));
+    return `await page.request.${method}(${params.join(', ')});`;
+  }
+
+  private prettyPrintObject(obj: any, indent = 2, level = 0): string {
+    // Handle null and undefined
+    if (obj === null)
+      return 'null';
+    if (obj === undefined)
+      return 'undefined';
+
+    // Handle primitive types
+    if (typeof obj !== 'object') {
+      if (typeof obj === 'string')
+        return this.stringLiteral(obj);
+      return String(obj);
+    }
+
+    // Handle arrays
+    if (Array.isArray(obj)) {
+      if (obj.length === 0)
+        return '[]';
+      const spaces = ' '.repeat(level * indent);
+      const nextSpaces = ' '.repeat((level + 1) * indent);
+
+      const items = obj.map(item =>
+        `${nextSpaces}${this.prettyPrintObject(item, indent, level + 1)}`
+      ).join(',\n');
+
+      return `[\n${items}\n${spaces}]`;
+    }
+
+    // Handle regular objects
+    if (Object.keys(obj).length === 0)
+      return '{}';
+    const spaces = ' '.repeat(level * indent);
+    const nextSpaces = ' '.repeat((level + 1) * indent);
+
+    const entries = Object.entries(obj).map(([key, value]) => {
+      const formattedValue = this.prettyPrintObject(value, indent, level + 1);
+      // Handle keys that need quotes
+      const formattedKey = /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key) ?
+        key :
+        this.stringLiteral(key);
+      return `${nextSpaces}${formattedKey}: ${formattedValue}`;
+    }).join(',\n');
+
+    return `{\n${entries}\n${spaces}}`;
+  }
+
+  private stringLiteral(v: string): string {
+    v = v.replace(/\\/g, '\\\\').replace(/'/g, '\\\'');
```

Comment:
This does not escape control characters like \n, \r, \t. `JSON.stringify(v)` would do a better job.

Reply:
The problem with `JSON.stringify` is that it outputs double-quotes, and I want the outputted code to conform to the style we have in the rest of the product, which uses single quotes.

> This does not escape control characters like \n, \r, \t.

Good catch! I'll address that.

Reply:
I've made it detect control characters and output them unescaped into source - I think the resulting code is easier to reason about than having a long string of escaped control characters.